pkg/providers/aws/aws_provider.go

```diff
@@ -0,0 +1,230 @@
+}
+
+type awsProvider struct {
+	config  awsProviderConfig
+	logger  logr.Logger
+	handler AWSHandlerInterface
+}
+
+func AWSProviderFactory(providerInfo string, secretData map[string][]byte, logger logr.Logger) (common.Provider, error) {
+	config := awsProviderConfig{
+		MachineSpec: &MachineSpec{
+			Regions: []RegionSpec{
+				{
+					Name: defaultRegion,
+					Instances: []InstanceSpec{
+						{Type: defualtInstanceType, AMIID: defaultAMI},
+					},
+				},
+			},
+			Device: BlockDeviceSpec{
+				DeviceName: defaultDeviceName,
+				DeviceSize: defaultDeviceSize,
+				DeviceType: defaultDeviceType,
+			},
+		},
+	}
+
+	if configJson, ok := secretData["config"]; ok {
+		if err := json.Unmarshal(configJson, &config); err != nil {
+			return nil, fmt.Errorf("error in provider config json: %w", err)
+		}
+	}
+
+	if config.UserData != "" {
+		// Provided userdata from secret should be base64 encoded
+		decodedBytes, err := base64.StdEncoding.DecodeString(config.UserData)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding userdata: %w", err)
+		}
+
+		config.UserData = string(decodedBytes)
+	}
+
+	awsHandler, err := NewAWSHandler(config.AccessKey, config.SecretAccessKey, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed creating AWS handler")
+	}
+
+	return &awsProvider{
+		config:  config,
+		logger:  logger,
+		handler: awsHandler,
+	}, nil
+}
+
+func (p *awsProvider) Acquire(poolSize int, poolName string, poolType string) (common.Resource, error) {
+	ctx := context.Background()
+
+	total := 0
+	for _, regionSpec := range p.config.MachineSpec.Regions {
+		count, err := p.handler.GetTotalAmountOfPoolInstancesInRegion(
+			ctx, &PoolFilterParams{Region: regionSpec.Name, PoolName: poolName},
+		)
+		if err != nil {
+			return common.Resource{}, errors.Wrapf(
+				err, "failed counting the total amount of instances in pool %s in region %s", poolName, regionSpec.Name,
+			)
+		}
+		total += lo.FromPtr(count)
+	}
+
+	if total >= poolSize {
+		return common.Resource{}, fmt.Errorf(
+			"refusing to create instance, already have %d and pool size is %d", total, poolSize,
+		)
+	}
+
+	var lastErr error
+	for _, regionSpec := range p.config.MachineSpec.Regions {
+		for _, instanceSpec := range regionSpec.Instances {
+			id, err := p.handler.RunInstanceInRegion(
+				ctx,
+				&RunInstanceParams{
+					Region:          regionSpec.Name,
+					InstanceType:    instanceSpec.Type,
+					AMI:             instanceSpec.AMIID,
+					KeyPair:         regionSpec.KeyPairName,
+					SecurityGroupID: regionSpec.SecurityGroupID,
+					SubnetID:        regionSpec.SubnetID,
+					UserData:        p.config.UserData,
+					PoolName:        poolName,
+					Device: BlockDeviceSpec{
+						DeviceName: p.config.MachineSpec.Device.DeviceName,
+						DeviceSize: p.config.MachineSpec.Device.DeviceSize,
+						DeviceType: p.config.MachineSpec.Device.DeviceType,
+					},
+				},
+			)
+			if err != nil {
+				p.logger.Error(err, "error running instance")
```

Comment:
some additional details could be helpful here for the troubleshooting

Reply:
All the details are coming from the error returned by `RunInstanceInRegion`

Reply:
Then it's fine, I wasn't sure if the err contained enough data